### PR TITLE
CRM: Add WooSync status to avoid creating invoices and transactions

### PR DESCRIPTION
**Pull request by gogdzl**
Fixes https://github.com/Automattic/zero-bs-crm/issues/3463

## Proposed changes:

* This PR adds a status mapping to WooSync to avoid creating invoices and transactions in Jetpack CRM from WooCommerce

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
https://github.com/Automattic/zero-bs-crm/issues/3463

## Does this pull request change what data or activity we track or use?
No.

## Testing instructions:
**Initial Setup and Verification of WooSync:**

1. **Verify WooSync Functionality:**
   - Navigate to the WooSync hub.
   - Confirm that the status displays `Status: Sync Completed`.

2. **Enable Invoice Generation:**
   - If invoice generation is not active, switch to the `trunk` branch and activate the feature at `/wp-admin/admin.php?page=zerobscrm-plugin-settings&tab=woosync`.

3. **Create a WooCommerce Order:**
   - Place a new order in WooCommerce.

4. **Verify Transaction and Invoice Creation:**
   - Ensure both the transaction and invoice are correctly generated.

**Testing the New PR Setting:**

1. **Switch to the PR Branch:**
   - Change to the branch `update/crm/3463-woosync-avoid-creating-invoices-and-transactions-for-woo-draft`.

2. **Configure WooCommerce Settings:**
   - Access Woo settings at `/wp-admin/admin.php?page=zerobscrm-plugin-settings&tab=woosync`.
   - Change the status mapping from one of WooCommerce's orders for at least one Invoice and one Transaction status to `Do not Create`.

3. **Add Another WooCommerce Order:**
   - Place a new order with the new mapping and confirm that WooSync is not generating Invoices when the mapping is set to `Do not Create`.
   - Place a new order with the new mapping and confirm that WooSync is not generating Transactions when the mapping is set to `Do not Create`.

## Documentantion changes

We will need to update our CRM Knowledge Base to reflect this change. It is necessary to update our WooSync page to inform users that we now offer a new setting and explain how it works.

Pinging @StefMattana 

image

### CODE REVIEW

File: projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php [hunk 1]
```diff
@@ -935,8 +935,8 @@ public function import_crm_object_data( $crm_object_data ) {
 
 		}
 
-		// Add/update invoice (if enabled) (previously `add_or_update_invoice`)
-		if ( $settings['wcinv'] == 1 ) {
+		// Add/update invoice (if enabled) (previously `add_or_update_invoice`), while checking for a 'Do not create' status to avoid creating this invoice if the mapping doesn't allow it
+		if ( $settings['wcinv'] == 1 && isset( $crm_object_data['invoice'] ) && isset( $crm_object_data['invoice']['status'] ) && $crm_object_data['invoice']['status'] !== WOOSYNC_DO_NOT_CREATE['id'] ) { // phpcs:ignore Universal.Operators.StrictComparisons.LooseEqual
```
Comment (cleacos) [inline projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php]: GH is complaining about something here...

File: projects/plugins/crm/modules/woo-sync/jpcrm-woo-sync-init.php [hunk 2]
```diff
@@ -243,6 +243,17 @@ function jpcrm_woosync_is_hpos_enabled() {
 	return false;
 }
 
+// If we ever have more WooSync constants we should create a separate php file.
+if ( ! defined( 'WOOSYNC_DO_NOT_CREATE' ) ) {
+	define(
+		'WOOSYNC_DO_NOT_CREATE',
```
Comment (cleacos) [inline projects/plugins/crm/modules/woo-sync/jpcrm-woo-sync-init.php]: For this global definition I'd add the `JPCRM_` prefix => `JPCRM_WOOSYNC_DO_NOT_CREATE`

**Review by cleacos — DISMISSED** [review]
It works well for me. I'd add the `JPCRM_` prefix to the `WOOSYNC_DO_NOT_CREATE` definition => `JPCRM_WOOSYNC_DO_NOT_CREATE`

Other than that, LGTM!

**Comment by cleacos** [comment]
As a side note, if an invoice is created because the order is different to Draft, if we set the order a Draft, the invoice/transaction doesn't change its status. 

For me that behavior is OK, once the invoice is created, you cannot go back automatically if the order is set as Draft later. But maybe I'd say that, if the invoice/transaction is already created and the new status should be Draft, we have to switch to that status.

**Comment by gogdzl** [comment]
> As a side note, if an invoice is created because the order is different to Draft, if we set the order a Draft, the invoice/transaction doesn't change its status.

That is a known problem, we sync only existing transactions, if a transactions gets deleted we don't record that.

File: projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php [hunk 3]
```diff
@@ -935,8 +935,8 @@ public function import_crm_object_data( $crm_object_data ) {
 
 		}
 
-		// Add/update invoice (if enabled) (previously `add_or_update_invoice`)
-		if ( $settings['wcinv'] == 1 ) {
+		// Add/update invoice (if enabled) (previously `add_or_update_invoice`), while checking for a 'Do not create' status to avoid creating this invoice if the mapping doesn't allow it
+		if ( $settings['wcinv'] == 1 && isset( $crm_object_data['invoice'] ) && isset( $crm_object_data['invoice']['status'] ) && $crm_object_data['invoice']['status'] !== JPCRM_WOOSYNC_DO_NOT_CREATE['id'] ) { // phpcs:ignore Universal.Operators.StrictComparisons.LooseEqual
```
Comment (cleacos) [inline projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php:940]: I don't know why GH is still complaining here...

**Review by cleacos — DISMISSED** [review]
Re-approved it...

A test is failing

**Comment by cleacos** [comment]
> > As a side note, if an invoice is created because the order is different to Draft, if we set the order a Draft, the invoice/transaction doesn't change its status.

> That is a known problem, we sync only existing transactions, if a transactions gets deleted we don't record that.

I mean, the user creates an Order in a valid status, let's say completed. JPCRM will generate the invoice as paid, but the user takes that order and changes it to Draft; in JPCRM, the invoice status will not change to Draft, it will leave the invoice as Paid. But it's a side note about this edge case not covered by this PR, not a blocker.

**Review by cleacos — APPROVED** [review]
👍

**Pull request merged**